src/mp_integer.hpp

```diff
@@ -844,6 +865,54 @@ struct static_integer
 		}
 		clear_extra_bits();
 	}
+	// lshift by n bits.
+	int lshift(unsigned long n)
+	{
+		using size_type = typename limbs_type::size_type;
+		if (n == 0) {
+			return 0;
+		} else if (n <= limb_bits) {
+			if (m_limbs[1u] >= (limb_t(1) << (limb_bits - n))) {
+				return 1;
+			}
+			// Shift both limbs.
+			const limb_t lo = m_limbs[0u] << n;
```

Comment:
Aren't we risking undefined behaviour here? If ``n == limb_bits`` and ``limb_bits`` is exactly the bit width of ``limb_t`` (as it is often the case), I am afraid we are shifting too much. Quoting from the standard:

> In any case, if the value of the right operand is negative or is greater or equal to the number of bits in the promoted left operand, the behavior is undefined.